src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.Collector.cs

```diff
@@ -0,0 +1,459 @@
+
+            /// <summary>
+            /// Trivial reject for fields that can't be upgraded in order to avoid wasted work.
+            /// </summary>
+            private static bool CanUpgrade(IFieldSymbol fieldSym)
+                => fieldSym.DeclaredAccessibility == Accessibility.Private;
+
+            private List<ITypeSymbol> GetValueTypes(IOperation op)
+            {
+                var values = new List<ITypeSymbol>();
+                GetValueTypes(values, op);
+                return values;
+            }
+
+            private void GetValueTypes(List<ITypeSymbol> values, IOperation op)
+            {
+                switch (op.Kind)
+                {
+                    case OperationKind.Literal:
+                        {
+                            var lit = (ILiteralOperation)op;
+                            if (lit.Type == null)
+                            {
+                                // use 'void' as a marker for a null assignment
+                                values.Add(Void!);
+                            }
+
+                            return;
+                        }
+
+                    case OperationKind.Conversion:
+                        {
+                            var convOp = (IConversionOperation)op;
+                            GetValueTypes(values, convOp.Operand);
+                            return;
+                        }
+
+                    case OperationKind.ConditionalAccess:
+                        {
+                            var condOp = (IConditionalAccessOperation)op;
+                            GetValueTypes(values, condOp.WhenNotNull);
+                            return;
+                        }
+
+                    case OperationKind.Conditional:
+                        {
+                            var condOp = (IConditionalOperation)op;
+                            GetValueTypes(values, condOp.WhenTrue);
+                            GetValueTypes(values, condOp.WhenFalse);
+                            return;
+                        }
+
+                    case OperationKind.Coalesce:
+                        {
+                            var colOp = (ICoalesceOperation)op;
+                            GetValueTypes(values, colOp.Value);
+                            GetValueTypes(values, colOp.WhenNull);
+                            return;
+                        }
+
+                    case OperationKind.Invocation:
+                    case OperationKind.ArrayElementReference:
+                    case OperationKind.ObjectCreation:
+                    case OperationKind.ParameterReference:
+                    case OperationKind.PropertyReference:
+                    case OperationKind.MethodReference:
+                    case OperationKind.LocalReference:
+                        {
+                            values.Add(op.Type!);
```

Comment:
Is null suppression safe here?


Here is a compiler test showing null `Type` for `IMethodReference` operation:

https://github.com/dotnet/roslyn/blob/ac92012c0e64c8114ecf0024c2bad2f63dcc7c6b/src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IDelegateCreationExpression.cs#L364-L396

Reply:
Good catch. Done.